add room type in store invite (#3288)
matrix-org/matrix-doc: proposals/3288-pass_room_type_in_3pid_invite.md

> # MSC3288: Make email invites space aware

Reviewer: It feels to soru like https://github.com/matrix-org/matrix-doc/pull/1840 should be figured out first?

Author: It's refering to #1772 immutable room type as it's the one used for spaces. Not sure if it's needed then

Currently when inviting via 3pid, the Identity Server is getting some information about the room,
like for example the room name and avatar as well as the inviter name.
This allows the identity server to generate a rich email to the invitee.

Now that the matrix spec supports spaces, it would be nice to also provide this information to the identity server
so that the email invite could be customized for spaces, current implementation would say wrongly that 
you are invited to a room when the room is actually a space.

The goal of this proposal is to make 3pid invites space aware.

## Proposal

Home servers should also send the `room_type` to the identity server when performing a third party invite (__Invitation storage__).

__Proposed change:__

Add a new `room_type` field in json body of `POST /_matrix/identity/v2/store-invite`

| Parameter | Type | Description |
|--|--|--|

> | room_type  | string  | The room type for the room to which the user is invited. This should be retrieved from the value of `type` in `m.room.create` event.

Reviewer 2: What do we do if there is no `type` field in the `m.room.create` event? I see two straightforward options, but I'm unsure which is expected:

* Do not include the `room_type` field.
* Default to `null` or `""`.

Author: Clarified that room_type should not be included in this case

````
POST /_matrix/identity/v2/store-invite HTTP/1.1
Content-Type: application/json

{
  "medium": "email",
  "address": "foo@example.com",
  "room_id": "!something:example.org",
  "sender": "@bob:example.com",
  "room_alias": "#somewhere:exmaple.org",
  "room_avatar_url": "mxc://example.org/s0meM3dia",
  "room_join_rules": "public",
  "room_name": "The Bob Project",
  "room_type": "m.space",
  "sender_display_name": "Bob Smith",
  "sender_avatar_url": "mxc://example.org/an0th3rM3dia"
}
````

The identity server could then use room type to customize the email depending on the room type.

## Potential issues

None.

## Security considerations

> None.

Reviewer 3: You might want to mention that #3173, #3266 and #2946 already deal with adding that field too (and a few more).

Reviewer 4: Deal with how? There's truly no worthwhile security concern with sending the room type to the identity server as far as I can tell.

## Unstable prefix

The following mapping will be used for identifiers in this MSC during development:

Proposed final identifier       | Purpose | Development identifier
------------------------------- | ------- | ----

> `room_type` | event type | `org.matrix.msc3288.room_type`

Reviewer 5: purpose isn't event_type, POST body field maybe?

Author: Fixed.

# MSC3288: Make email invites space aware

> Currently when inviting via 3pid, the Identity Server is getting some information about the room,

Reviewer 6: ```suggestion
Currently when inviting via 3pid, the Identity Server receives some information about the room,
```

like for example the room name and avatar as well as the inviter name.
This allows the identity server to generate a rich email to the invitee.

Now that the matrix spec supports spaces, it would be nice to also provide this information to the identity server

> so that the email invite could be customized for spaces, current implementation would say wrongly that 

Reviewer 6: ```suggestion
so that the email invite could be customized for spaces.  The current implementation would say wrongly that 
```

you are invited to a room when the room is actually a space.

The goal of this proposal is to make 3pid invites space aware.

## Proposal

> Home servers should also send the `room_type` to the identity server when performing a third party invite (__Invitation storage__).

Reviewer 4: ```suggestion
Homeservers should also send the `room_type` to the identity server when performing a third party invite (__Invitation storage__).
```

__Proposed change:__

> Add a new `room_type` field in json body of `POST /_matrix/identity/v2/store-invite`

Reviewer 6: A link to the existing API is super helpful:

```suggestion
Add a new `room_type` field in json body of [`POST /_matrix/identity/v2/store-invite`](https://matrix.org/docs/spec/identity_service/r0.3.0#post-matrix-identity-v2-store-invite):
```

| Parameter | Type | Description |
|--|--|--|

> | room_type  | string  | The room type for the room to which the user is invited. This should be retrieved from the value of `type` in `m.room.create` event. Do not include paremeter if not `type` not present in `m.room.create`.

Reviewer 4: ```suggestion
| room_type  | string  | The room type for the room to which the user is invited. This should be retrieved from the value of `type` in the `m.room.create` event's `content`. Do not include parameter if not `type` not present in `m.room.create`.
```

````
POST /_matrix/identity/v2/store-invite HTTP/1.1
Content-Type: application/json

{
  "medium": "email",
  "address": "foo@example.com",
  "room_id": "!something:example.org",
  "sender": "@bob:example.com",
  "room_alias": "#somewhere:exmaple.org",
  "room_avatar_url": "mxc://example.org/s0meM3dia",
  "room_join_rules": "public",
  "room_name": "The Bob Project",
  "room_type": "m.space",
  "sender_display_name": "Bob Smith",
  "sender_avatar_url": "mxc://example.org/an0th3rM3dia"
}
````

The identity server could then use room type to customize the email depending on the room type.

__Email Generation__

The link in the generated email should also pass over the `room_type` to clients ( like it is doing for 

> `inviter_name`, `room_name`, `room_avatar`)

Reviewer 7: I don't think any of this bit is specced currently...
